ts/components/Pinpad/index.tsx

```diff
@@ -0,0 +1,107 @@
+import { fromNullable } from "fp-ts/lib/Option";
+import { View } from "native-base";
+import * as React from "react";
+import { TextInput, TouchableOpacity } from "react-native";
+
+import { PinString } from "../../types/PinString";
+import { PIN_LENGTH } from "../../utils/constants";
+
+import { styles } from "./Pinpad.style";
+import { Baseline, Bullet } from "./Placeholders";
+
+const fucusElement = (el: TextInput) => el.focus();
+const blurElement = (el: TextInput) => el.blur();
+const current = (ref: React.RefObject<TextInput>) => ref.current;
+
+interface Props {
+  compareWithCode?: string;
+  inactiveColor: string;
+  activeColor: string;
+  onFulfill: (code: PinString, isValid: boolean) => void;
+}
+
+interface State {
+  value: PinString;
+}
+
+/**
+ * A customized CodeInput component.
+ */
+class Pinpad extends React.PureComponent<Props, State> {
+  private inputRef: React.RefObject<TextInput>;
+  private placeholderPositions: ReadonlyArray<undefined>;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      value: "" as PinString
+    };
+
+    this.inputRef = React.createRef();
+    this.placeholderPositions = [...new Array(PIN_LENGTH)];
+  }
+
+  public foldInputRef = (fn: (el: TextInput) => void) =>
```

Comment:
what does this do?

Reply:
When `this.inputRef.current` is `some`, map `fn` to it.
I used this to avoid using `if (this.inputRef && this.inputRef.current)` every time `focus` and `blur` methods of `inputRef` are invoked.